runtime/major_gc.c

```diff
@@ -1450,78 +1473,105 @@ void caml_finish_sweeping (void)
+  uintnat k = PTR_TO_PAGE(v);
+  uintnat flag = (uintnat)1 << PTR_TO_PAGE_OFFSET(v);
+  struct compressed_ptr_page *p = NULL;
+  int new_entry = 0;
+
+  value* amap_pos = caml_addrmap_insert_pos(amap, k);
+
+  if (*amap_pos == ADDRMAP_NOT_PRESENT) {
+    p = caml_stat_alloc_noexc(sizeof(struct compressed_ptr_page));
+    if (p == NULL)
+      caml_fatal_error("Unable to allocate for compressed entry address");
+    p->bitfield = p->count = 0;
+    new_entry = 1;
+    *amap_pos = (value)p;
   } else {
-    p = 0;
+    p = (struct compressed_ptr_page *)*amap_pos;
+  }
+
+  CAMLassert(v == (value)((k + PTR_TO_PAGE_OFFSET(v))<<3));
+
+  if (!(p->bitfield & flag)) {
+    p->bitfield |= flag;
+    p->count++;
   }
 
-  return p;
+  return new_entry;
 }
 
-static void mark_stack_prune (struct mark_stack* stk)
+static void mark_stack_prune(struct mark_stack* stk)
 {
-  int entry_idx, large_idx = 0;
-  mark_entry* mark_stack = stk->stack;
-
-  struct skiplist chunk_sklist = SKIPLIST_STATIC_INITIALIZER;
-  /* Insert used pools into skiplist */
-  for(entry_idx = 0; entry_idx < stk->count; entry_idx++){
-    mark_entry me = mark_stack[entry_idx];
-    struct pool* pool = caml_pool_of_shared_block(Val_op(me.start));
-    if (!pool) {
-      // This could be a large allocation - which is off-heap. Hold on to it.
-      mark_stack[large_idx++] = me;
-      continue;
-    }
-    caml_skiplist_insert(&chunk_sklist, (uintnat)pool,
-    (uintnat)pool + sizeof(pool));
+  /* copy out old compressed entries */
+  uintnat old_compressed_entries = 0;
+  struct addrmap new_compressed_stack = ADDRMAP_INIT;
+  addrmap_iterator it;
+  for (it = stk->compressed_stack_iter;
+       caml_addrmap_iter_ok(&stk->compressed_stack, it);
+       it = caml_addrmap_next(&stk->compressed_stack, it)) {
+    value k = caml_addrmap_iter_key(&stk->compressed_stack, it);
+    value v = caml_addrmap_iter_value(&stk->compressed_stack, it);
+    caml_addrmap_insert(&new_compressed_stack, k, v);
+    ++old_compressed_entries;
+  }
+  if (old_compressed_entries > 0) {
+    caml_gc_log("Preserved %"ARCH_INTNAT_PRINTF_FORMAT "d compressed entries",
+                old_compressed_entries);
   }
 
-  /* Traverse through entire skiplist and put it into pools to rescan */
-  FOREACH_SKIPLIST_ELEMENT(e, &chunk_sklist, {
-    if(Caml_state->pools_to_rescan_len == Caml_state->pools_to_rescan_count){
-      Caml_state->pools_to_rescan_len =
-        Caml_state->pools_to_rescan_len * 2 + 128;
-
-      Caml_state->pools_to_rescan =
-        caml_stat_resize(
-          Caml_state->pools_to_rescan,
-          Caml_state->pools_to_rescan_len * sizeof(struct pool *));
+  /* scan mark stack and determine page densities */
+  int i;
+  uintnat new_stk_count = 0, compressed_entries = 0, total_words = 0;
+  for (i=0; i < stk->count; i++) {
+    mark_entry me = stk->stack[i];
+    total_words += me.end - me.start;
+    if (me.end - me.start > BITS_PER_WORD) {
```

Comment:
I've been pondering if there's a heap structure composed of 65 field blocks that would repeatedly result in spilling the mark stack.

Reply:
It should be OK unless I made a mistake in the following.

Let us consider a heap entirely composed of blocks of sizes 65 and 1, all alive. I think this is the worst case because it's the minimum size of either kind of blocks, so it'll give the largest number of blocks, hence the most stack use. Note I didn't look for a proof that this is the worst case.

we define:
- H = heap size
- S = stack size
- x = number of size-65 blocks
- y = number of size-1 blocks

Then we have:
- H = 66x + 2y  (accounting for headers)

thus
- s = N/2 - 33x

The large blocks will stay in the stack while we scan the small blocks, so the useful stack size is S - x and the number of compressions will be y / (S - x) = (H/2 - 33x) / (S - x)

Let us study the function f(x) = H/2 - 33x / (S - x) between x = 0 and x = H/66 (where y = 0).

If x gets up to S, then everything breaks because the stack gets full of large blocks and we cannot make progress, so we have the constraint:
```
  S > H/66
```

The derivative of f is f'(x) = (H/2 - 33S) / (S - x)^2. It has the sign of H/2 - 33S or H/66 - S, so it's negative everywhere (because of our constraint). Thus f is decreasing on the whole range [0, H/66]. Its maximum is at x = 0, where its value is H/2S.

Hence we cannot have more than
```
Cmax = H/2S
```
stack compressions.

The code increases the stack as long as it is smaller than H/32, so we have S >= H/32 (satisfying our constraint) thus H <= 32S and **Cmax <= 16**. This is reassuring.

Now for the case of 32-bit machines: in the above replace 66 by 34 and 33 by 17, you get the constraint S > H/34, still satisfied by the code. Cmax is unchanged (it occurs at x=0, when there are no large blocks, so it makes sense that their size is irrelevant).

Reply:
How about keeping track of how many large blocks are in the stack, and ignoring them in the stack size calculation? Then your maximum stack size is local_heap/32 + global_heap/65, where the global_heap/65 term is shared between all threads, so the maximum size of all the stacks together is sum(local_heap)/32 + global_heap/65. Since sum(local_heap) is global_heap, that gives you a total stack size of about global_heap/20, which gets automatically distributed between the domains. Maybe change the numbers to get it closer to global_heap/32.

Reply:
I like your idea better - it will also make explicit that there is always space for the large blocks we keep. I will have a go at that.

Reply:
I have implemented this. I did not tweak the constants (but we could). Right now the bound is:
 - 64bits: global_heap/32 + global_heap/65 or ~0.0466 of global_heap
 - 32bits: global_heap/32 + global_heap/33 or ~0.0616 of global_heap
If someone feels strongly about tweaking the constants, I'm happy to do it.